src/main/java/duckling/requests/Request.java

```diff
@@ -0,0 +1,104 @@
+package duckling.requests;
+
+import duckling.Server;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Hashtable;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class Request {
+    protected boolean acceptingBody = false;
+    protected InitialRequest initialRequest;
+    protected Hashtable<String, String> headers = new Hashtable<>();
+    protected ArrayList<String> body = new ArrayList<>();
+    private String root;
+
+    public Request(String root) {
+        this.root = root;
+    }
+
+    public Request() {
+        this(".");
+    }
+
+    public String getBody() {
+        return body.stream().collect(Collectors.joining(Server.CLRF));
+    }
+
+    public String getHeader(String key) {
+        return headers.get(key);
+    }
+
+    public boolean isAcceptingBody() {
+        return acceptingBody;
+    }
+
+    public void setAcceptingBody() {
+        this.acceptingBody = true;
+    }
+
+    public void add(String line) {
+        if (line.length() == 0) setAcceptingBody();
+        if (this.initialRequest == null && !acceptingBody) setInitialRequest(line);
+
+        if (acceptingBody && line.length() != 0) {
+            body.add(line);
+        } else {
+            setHeaderPair(line);
+        }
+    }
+
+    private void setHeaderPair(String line) {
+        String[] headerPair = line.split(": ");
+
+        try {
+            headers.put(headerPair[0], headerPair[1]);
+        } catch (ArrayIndexOutOfBoundsException exception) {
+        }
+    }
+
+    private void setInitialRequest(String initialRequest) {
+        this.initialRequest = new InitialRequest(initialRequest);
+    }
+
+    public String initialRequestString() {
+        return Stream.of(
+            initialRequest.getMethod(),
+            initialRequest.getPath(),
+            initialRequest.getProtocol()
+        ).collect(Collectors.joining(" "));
+    }
+
+    public String getMethod() {
+        return initialRequest.getMethod();
+    }
+
+    public String getPath() {
+        return initialRequest.getPath();
+    }
+
+    public String getProtocol() {
+        return initialRequest.getProtocol();
+    }
+
+    public File getFile() {
+        return new File(fullFilePath());
+    }
+
+    public String fullFilePath() {
+        return root + getPath();
+    }
+
+    public boolean equals(Request other) {
+        return initialRequest.equals(other.initialRequest)
+                && body.equals(other.body)
+                && headers.equals(other.headers);
+    }
```

Comment:
Rule of thumbs is that when you override `equals`, you should also override `hashCode` so that things like using objects as HashMap/HashSet keys work as expected. For example, if 2 objects have different `hashCode` values but are equal, hash lookups aren't guaranteed to be able to find find them (see http://stackoverflow.com/questions/27581/what-issues-should-be-considered-when-overriding-equals-and-hashcode-in-java).

Also: the Java `equals` method should `@Override` the one from `java.lang.Object`, which requires some additional null & class checks for full correctness. There are some things in Java-land that will use the `equals(Object other)` version under the hood, which will come back and bite you later. IntelliJ can also auto-generate some of the boilerplate for you there.